### ISSUE TITLE
connections to k8s services from inside the k8s cluster randomly fail

### ISSUE THREAD

**Issue by axinojolais** [juju-solutions/bundle-canonical-kubernetes#166]
Hi,

This is happening on a k8s cluster with 3 workersdeployed on OpenStack, with default networking options.

I deployed the microbot application (using the juju action), it appears as below : 

```
$ kubectl get svc,ep,po -o wide|grep microb
svc/microbot               10.152.183.196   <none>        80/TCP    1d        app=microbot

ep/microbot               10.1.62.3:80,10.1.92.6:80,10.1.92.7:80   1d

po/microbot-1855935831-68dlh        1/1       Running   0          1d        10.1.62.3      juju-ffddf0-juju2-is-kubernetes-39-7
po/microbot-1855935831-74tds        1/1       Running   0          1d        10.1.92.6      juju-ffddf0-juju2-is-kubernetes-39-1
po/microbot-1855935831-w36ls        1/1       Running   0          2h        10.1.92.7      juju-ffddf0-juju2-is-kubernetes-39-1
```

However, when I connect on a kubernetes-worker unit, and try to connect to 10.152.183.196 port 80 from that unit, it randomly works : 
```
$ while :; do nc -w1 -z -v 10.152.183.196 80; sleep 1; done
nc: connect to 10.152.183.196 port 80 (tcp) timed out: Operation now in progress
nc: connect to 10.152.183.196 port 80 (tcp) timed out: Operation now in progress
Connection to 10.152.183.196 80 port [tcp/http] succeeded!
Connection to 10.152.183.196 80 port [tcp/http] succeeded!
nc: connect to 10.152.183.196 port 80 (tcp) timed out: Operation now in progress
nc: connect to 10.152.183.196 port 80 (tcp) timed out: Operation now in progress
nc: connect to 10.152.183.196 port 80 (tcp) timed out: Operation now in progress
Connection to 10.152.183.196 80 port [tcp/http] succeeded!
Connection to 10.152.183.196 80 port [tcp/http] succeeded!
Connection to 10.152.183.196 80 port [tcp/http] succeeded!
nc: connect to 10.152.183.196 port 80 (tcp) timed out: Operation now in progress
nc: connect to 10.152.183.196 port 80 (tcp) timed out: Operation now in progress
nc: connect to 10.152.183.196 port 80 (tcp) timed out: Operation now in progress
nc: connect to 10.152.183.196 port 80 (tcp) timed out: Operation now in progress
Connection to 10.152.183.196 80 port [tcp/http] succeeded!
Connection to 10.152.183.196 80 port [tcp/http] succeeded!
nc: connect to 10.152.183.196 port 80 (tcp) timed out: Operation now in progress
Connection to 10.152.183.196 80 port [tcp/http] succeeded!
nc: connect to 10.152.183.196 port 80 (tcp) timed out: Operation now in progress
Connection to 10.152.183.196 80 port [tcp/http] succeeded!
nc: connect to 10.152.183.196 port 80 (tcp) timed out: Operation now in progress
nc: connect to 10.152.183.196 port 80 (tcp) timed out: Operation now in progress
Connection to 10.152.183.196 80 port [tcp/http] succeeded!
nc: connect to 10.152.183.196 port 80 (tcp) timed out: Operation now in progress
nc: connect to 10.152.183.196 port 80 (tcp) timed out: Operation now in progress
nc: connect to 10.152.183.196 port 80 (tcp) timed out: Operation now in progress
Connection to 10.152.183.196 80 port [tcp/http] succeeded!
nc: connect to 10.152.183.196 port 80 (tcp) timed out: Operation now in progress
nc: connect to 10.152.183.196 port 80 (tcp) timed out: Operation now in progress
nc: connect to 10.152.183.196 port 80 (tcp) timed out: Operation now in progress
nc: connect to 10.152.183.196 port 80 (tcp) timed out: Operation now in progress
nc: connect to 10.152.183.196 port 80 (tcp) timed out: Operation now in progress
^C
```

I found out that the connection succeeds only when it's redirected to the pod on the local worker. Whenever it's sent to another worker, the packet is treated as a martian : 
```
IPv4: martian source 10.1.92.7 from 10.25.62.148, on dev flannel.1
```

The networking configuration is the following : 
* Source worker : 
```
$ ip r
default via 10.25.60.1 dev ens3 
10.1.0.0/16 dev flannel.1  proto kernel  scope link  src 10.1.62.0 
10.1.62.0/24 dev cni0  proto kernel  scope link  src 10.1.62.1 
10.25.60.0/22 dev ens3  proto kernel  scope link  src 10.25.62.148 
172.17.0.0/16 dev docker0  proto kernel  scope link  src 172.17.0.1 linkdown 
$ ip -4 a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
2: ens3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    inet 10.25.62.148/22 brd 10.25.63.255 scope global ens3
       valid_lft forever preferred_lft forever
4: docker0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN group default 
    inet 172.17.0.1/16 scope global docker0
       valid_lft forever preferred_lft forever
5: flannel.1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1450 qdisc noqueue state UNKNOWN group default 
    inet 10.1.62.0/16 scope global flannel.1
       valid_lft forever preferred_lft forever
6: cni0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1450 qdisc noqueue state UP group default qlen 1000
    inet 10.1.62.1/24 scope global cni0
       valid_lft forever preferred_lft forever
```

* Destination worker
```
$ ip r
default via 10.25.60.1 dev ens3 
10.1.0.0/16 dev flannel.1  proto kernel  scope link  src 10.1.92.0 
10.1.92.0/24 dev cni0  proto kernel  scope link  src 10.1.92.1 
10.25.60.0/22 dev ens3  proto kernel  scope link  src 10.25.62.142 
172.17.0.0/16 dev docker0  proto kernel  scope link  src 172.17.0.1 linkdown 
$ ip -4 a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
2: ens3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    inet 10.25.62.142/22 brd 10.25.63.255 scope global ens3
       valid_lft forever preferred_lft forever
4: docker0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN group default 
    inet 172.17.0.1/16 scope global docker0
       valid_lft forever preferred_lft forever
5: flannel.1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1450 qdisc noqueue state UNKNOWN group default 
    inet 10.1.92.0/16 scope global flannel.1
       valid_lft forever preferred_lft forever
6: cni0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1450 qdisc noqueue state UP group default qlen 1000
    inet 10.1.92.1/24 scope global cni0
       valid_lft forever preferred_lft forever
```

Packet arriving on the destination worker, as seen by tcpdump, on the interface flannel.1 : 
12:58:34.945922 IP 10.25.62.148.46988 > 10.1.92.7.80: Flags [S], seq 4133341387, win 29200, options [mss 1460,sackOK,TS val 45071626 ecr 0,nop,wscale 7], length 0

This is martian because 10.25.62.148 isn't on flannel.1 - for the kernel, a packet with such a source IP can only arrive from ens3. 10.25.62.148 is the IP of the OpenStack instance.

I'm not sure what the root cause if for this though. Can you reproduce this behaviour ? I can provide more logs / run more tests if necessary.

Thanks !

**Comment by jacekn** [comment]
A workaround is to add static route. In my case:

source worker: ens3 10.25.61.29; flannel.1 10.1.38.0/16

When I try testing like this:

    nc -w1 -z -v 10.152.183.60 80

This is what I see in "tcpdump -i flannel.1 -lnn" output on the other end (I removed part of the output for readability:

    15:11:11.399747 IP 10.25.61.29.47358 > 10.1.39.16.80: Flags [S]

And connection fails. If I run this on the source worker:

    sudo ip r add 10.152.183.60/32 dev flannel.1 src 10.1.38.0

It works fine and this is tcpdump output on the other end (note that source IP is now 10.1.38.0 as opposed to 10.25.61.29):
````
15:15:25.467056 IP 10.1.38.0.54928 > 10.1.39.16.80: Flags [S], seq 2248059407, win 28200, options [mss 1410,sackOK,TS val 64215687 ecr 0,nop,wscale 7], length 0
15:15:25.467163 IP 10.1.39.16.80 > 10.1.38.0.54928: Flags [S.], seq 3892254272, ack 2248059408, win 27960, options [mss 1410,sackOK,TS val 64211608 ecr 64215687,nop,wscale 7], length 0
15:15:25.468563 IP 10.1.38.0.54928 > 10.1.39.16.80: Flags [.], ack 1, win 221, options [nop,nop,TS val 64215688 ecr 64211608], length 0
15:15:25.468848 IP 10.1.38.0.54928 > 10.1.39.16.80: Flags [F.], seq 1, ack 1, win 221, options [nop,nop,TS val 64215688 ecr 64211608], length 0
15:15:25.468945 IP 10.1.39.16.80 > 10.1.38.0.54928: Flags [F.], seq 1, ack 2, win 219, options [nop,nop,TS val 64211608 ecr 64215688], length 0
15:15:25.469886 IP 10.1.38.0.54928 > 10.1.39.16.80: Flags [.], ack 2, win 221, options [nop,nop,TS val 64215688 ecr 64211608], length 0
````

**Comment by Cynerva** [comment]
We should try passing `--cluster-cidr` to kube-proxy

**Comment by jacekn** [comment]
@Cynerva yes I think you're right. I've just tested and it things started working for me.

**Comment by wwwtyro** [comment]
Alright, we've got a fix in that should be in the 1.6 release. Thanks for bringing this to our attention, @axinojolais
